### DISCUSSION TITLE
[BUG] Incorrect Display on Screen after Number Rounding

### DISCUSSION THREAD

**Discussion by kamitojaeger** [cc-tweaked/CC-Tweaked#1444] [Q&A]
### Minecraft Version

1.18.x

### Version

1.101.2

### Details

I have encountered a bug while using the CC Computer mod. I have written a program that monitors the total quantity of items in a chest and continuously prints it on a monitor screen at a fixed position. The code implementation is as follows:

lua
Copy code
local chest = peripheral.wrap("back")
local monitor = peripheral.wrap("monitor")
local sum = 0

while true do
  for slot, item in pairs(chest.list()) do
    sum = sum + item.count
  end

  monitor.setCursorPos(1, 5)
  monitor.write()
  monitor.write("Gun_1:")
  monitor.write(sum)
  sum = 0
  sleep(0.1)
end
In the code above, I used monitor.write() to clear the content at position (1,5) on the screen and then printed the subsequent "Gun_1:" and "sum" values. At this point, if I don't round the variable sum and display it as a decimal, the monitor display works correctly.

However, as soon as I round the sum variable to an integer using any method, and then display the rounded value on the monitor, the display becomes incorrect.

For example, when the displayed value transitions from a four-digit number to a three-digit number or from a three-digit number to a two-digit number, the following issue occurs:

Original: 128

After removing 64 items, the displayed content should become 64. However, in reality, the displayed content becomes 648.

The last digit of the previous three-digit number does not disappear from the screen. If I break and replace the monitor block, forming a complete monitor screen again, the display content becomes normal, showing 64. It seems that there is a problem with the screen refresh.

**Comment by SquidDev** [comment]
When you call `monitor.write`, it just writes the text you specify - it doesn't clear any of the other text. This means when you write `64`, you just replace _just_ the first two characters of `128`, hence the `648`.

You can add a call to `monitor.clearLine()` after `monitor.setCursorPos(1, 5)`, which should fix your problem.

**Reply by kamitojaeger** [reply]
I know I can use monitor.clearLine(x,y) to clear the line. But because I also display some other texts at the same line. I only want to clear part of the line.

that's why I use
monitor.write()
to try to clear part of the line

If [monitor.write()] couldn't clear any of the text
then the code above should have made the monitor displayed: 
Gun_L:sumGun_L:sumGun_L:sumGun_L:sumGun_L:sumGun_L:sum......
at the same Line

But it did only desplay Gun_L: sum
with the variable sum changing to the items in the chest

And about "This means when you write 64, you just replace just the first two characters of 128, hence the 648."  
if  i don't round the number, it will display just fine.
from 128 directly to 64
but if i round the number, it will display wrong
from 128 to 648

I also tried using this:

monitor.setCursorPos(1,5)
monitor.write("")
monitor.setCursorPos(2,5)
monitor.write("")
monitor.setCursorPos(3,5)
monitor.write("")
monitor.setCursorPos(4,5)
monitor.write("")
monitor.setCursorPos(5,5)
monitor.write("")
monitor.setCursorPos(6,5)
monitor.write("")
monitor.setCursorPos(7,5)
monitor.write("")
monitor.setCursorPos(8,5)
monitor.write("")
monitor.setCursorPos(9,5)
monitor.write("")

but i still couldn't clear part of the line

**Reply by SquidDev** [reply]
Just a couple of quick questions:

 - What code are you using to round? I'd expect` sum` to be an integer already (as there won't be any items with half a stack), so surprised that makes a difference.
 - Just to confirm, are you calling `monitor.write("")` with an empty string, or are there spaces in your original code that GitHub has trimmed (you can use triple backticks before and after code to make it monospace: ` ``` `)? Calling `.write` with an empty string won't do anything at all.

**Reply by kamitojaeger** [reply]
1. I'm also confused why the sum has a ".00" behind it even if it is an integer.
2. The sum is displayed like:{"0.000", "64.00", "128.00"} if i don't sum it.
3. I have tried [string.format("%.0f",sum)]  and [tostring(math.floor(sum))] to round the sum.
4. I didn't add spaces when calling [monitor.write("")] because the spaces are considered as string and will occupy part of the line.

**Reply by SquidDev** [reply]
> I'm also confused why the sum has a ".00" behind it even if it is an integer.

Yeah, this is caused by some wonky conversion on the Java side. It's been bugging me for a while - I've opened #1445 to address that. It's not actually the rounding which matters here, it's that the argument passed to `write` is a number rather than a string. Doing `monitor.write(tostring(sum))` will also work.

> And if Calling `.write` with an empty string won't do anything at all, how can it still clear the places when i didn't round the sum tho

It doesn't. I suspect you're seeing something else going on here.

**Comment by kamitojaeger** [comment]
And if Calling .write with an empty string won't do anything at all, how can it still clear the places when i didn't round the sum tho

**Reply by Lupus590** [reply]
Write spaces instead of empty strings. Write will only change as many characters as you provide, if you give no characters (i.e. an empty string) then it won't do anything.
